File: chirp2bc125.go
package main

import (
	"encoding/csv"
	"flag"
	"fmt"
	"os"
	"strconv"
)

func main() {
	inputFileName := flag.String("input", "", "Path to the input Chirp CSV file")
	outputFileName := flag.String("output", "", "Path to the output bc125 CSV file")
	startNumber := flag.Int("start-number", 1, "Starting number for the first column")
	flag.Parse()

	if *inputFileName == "" || *outputFileName == "" {
		fmt.Println("Both input and output file paths are required.")
		return
	}

	inputFile, err := os.Open(*inputFileName)
	if err != nil {
		fmt.Printf("Error opening input file: %v\n", err)
		return
	}
	defer inputFile.Close()

	outputFile, err := os.Create(*outputFileName)
	if err != nil {
		fmt.Printf("Error creating output file: %v\n", err)
		return
	}
	defer outputFile.Close()

	csvReader := csv.NewReader(inputFile)
	csvWriter := csv.NewWriter(outputFile)

	header := []string{"Channel", "Name", "Frequency", "Modulation", "CTCSS/DCS", "Delay", "Lockout", "Priority"}
	csvWriter.Write(header)

	i := *startNumber
	for {
		row, err := csvReader.Read()
		if err != nil {
			break
		}

		location := strconv.Itoa(i)
		ctcssDcs := convertCTCSSDcs(row[5], row[6])
		writeRowToCSV(csvWriter, location, row[1], row[2], row[10], ctcssDcs)

		i++
	}

	csvWriter.Flush()
	if err := csvWriter.Error(); err != nil {
		fmt.Printf("Error writing to output file: %v\n", err)
	}

	fmt.Println("Conversion completed.")
}

func convertCTCSSDcs(tone, rToneFreq string) string {
	if (tone == "Tone" || tone == "SQL") && rToneFreq != "" {
		f, err := strconv.ParseFloat(rToneFreq, 64)
		if err == nil {
			return fmt.Sprintf("%.1f Hz", f)
		}
	}
	return "none"
}

func writeRowToCSV(csvWriter *csv.Writer, location, name, frequency, mode, ctcssDcs string) {
	csvWriter.Write([]string{location, name, frequency, mode, ctcssDcs, "2", "no", "no"})
}

File: chirp2bc125.py
#!/usr/bin/env python3
import argparse
import csv


def convert_chirp_to_bc125(input_file, output_file, start_number):
    with open(input_file, 'r') as input_csv, open(output_file, 'w',
                                                  newline='') as output_csv:
        csv_reader = csv.DictReader(input_csv)
        fieldnames = ["Channel", "Name", "Frequency", "Modulation",
                      "CTCSS/DCS", "Delay", "Lockout", "Priority"]
        csv_writer = csv.DictWriter(output_csv, fieldnames=fieldnames)
        csv_writer.writeheader()

        for i, row in enumerate(csv_reader, start=start_number):
            ctcss_dcs = convert_ctcss_dcs(row['Tone'], row['rToneFreq'])
            write_row_to_csv(csv_writer, str(
                i), row['Name'], row['Frequency'], row['Mode'], ctcss_dcs)


def convert_ctcss_dcs(tone, r_tone_freq):
    if tone in ['Tone', 'SQL'] and r_tone_freq:
        return f'{float(r_tone_freq):.1f} Hz'
    return 'none'


def write_row_to_csv(csv_writer, location, name, frequency, mode, ctcss_dcs):
    csv_writer.writerow({
        "Channel": location,
        "Name": name,
        "Frequency": frequency,
        "Modulation": mode,
        "CTCSS/DCS": ctcss_dcs,
        "Delay": "2",
        "Lockout": "no",
        "Priority": "no"
    })


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Convert Chirp CSV format to bc125 CSV format")

    parser.add_argument("--input", dest="input_file",
                        required=True,
                        help="Path to the input Chirp CSV file")
    parser.add_argument("--output", dest="output_file",
                        required=True,
                        help="Path to the output bc125 CSV file")
    parser.add_argument("--start-number", type=int, default=1,
                        help="Starting number for the first column")

    args = parser.parse_args()

    convert_chirp_to_bc125(
        args.input_file, args.output_file, args.start_number)
